src/main/java/com/iota/iri/utils/dag/DAGUtils.java

```diff
@@ -0,0 +1,366 @@
+ * to be implemented over and over again in different parts of the code.
+ */
+public class DAGUtils {
+    /**
+     * Holds references to the singleton DAGUtils instances.
+     */
+    protected static HashMap<Object, DAGUtils> instances = new HashMap<>();
+
+    /**
+     * Holds a reference to the tangle instance which acts as an interface to the used database.
+     */
+    protected Tangle tangle;
+
+    /**
+     * This method allows us to retrieve the DAGUtils instance that corresponds to the given parameters.
+     *
+     * To save memory and to be able to use things like caches effectively, we do not allow to create multiple instances
+     * of this class for the same set of parameters, but create a singleton instance of it.
+     *
+     * @param tangle database interface used to retrieve data
+     * @return DAGUtils instance that allows to use the implemented methods.
+     */
+    public static DAGUtils get(Tangle tangle) {
+        DAGUtils instance = instances.get(tangle);
+        if(instance == null) {
+            instances.put(tangle, instance = new DAGUtils(tangle));
+        }
+
+        return instance;
+    }
+
+    /**
+     * Constructor of the class which allows to provide the required dependencies for the implemented methods.
+     *
+     * Since the constructor is protected we are not able to manually create instances of this class and have to
+     * retrieve them through their singleton accessor {@link #get}.
+     *
+     * @param tangle database interface used to retrieve data
+     */
+    protected DAGUtils(Tangle tangle) {
+        this.tangle = tangle;
+    }
+
+    //region TRAVERSE APPROVERS (BOTTOM -> TOP) ////////////////////////////////////////////////////////////////////////
+
+    /**
+     * This method provides a generic interface for traversing all approvers of a transaction up to a given point.
+     *
+     * To make the use of this method as easy as possible we provide a broad range of possible parameter signatures, so
+     * we do not have to manually convert between milestones, transactions and hashes (see other methods with the same
+     * name).
+     *
+     * It uses an non-recursive iterative algorithm that is able to handle huge chunks of the tangle without running out
+     * of memory. It creates a queue of transactions that are being examined and processes them one by one. As new
+     * approvers are found, they will be added to the queue and processed accordingly.
+     *
+     * Every found transaction is passed into the provided condition lambda, to determine if it still belongs to the
+     * desired set of transactions and only then will be passed on to the currentTransactionConsumer lambda.
+     *
+     * @param startingTransaction the starting point of the traversal (does not get passed to the consumer)
+     * @param condition a lambda function that allows to control how long the traversal should continue
```

Comment:
Can you please add to the documentation that you pass the current traversed transaction to the condition lambda?

Reply:
can do